src/coprocessor/dag/batch/executors/top_n_executor.rs

```diff
@@ -0,0 +1,722 @@
+            for (column_idx, column) in item.source_column_data.iter().enumerate() {
+                match column {
+                    LazyBatchColumn::Raw(src_value) => {
+                        result[column_idx]
+                            .mut_raw()
+                            .push(&src_value[item.row_index]);
+                    }
+                    LazyBatchColumn::Decoded(src_value) => {
+                        match_template_evaluable! {
+                            TT, match (src_value, result[column_idx].mut_decoded()) {
+                                (VectorValue::TT(src), VectorValue::TT(dest)) => {
+                                    dest.push(src[item.row_index].clone());
+                                },
+                                _ => unreachable!(),
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        result
+    }
+}
+
+impl<Src: BatchExecutor> BatchExecutor for BatchTopNExecutor<Src> {
+    #[inline]
+    fn schema(&self) -> &[FieldType] {
+        self.src.schema()
+    }
+
+    #[inline]
+    fn next_batch(&mut self, _scan_rows: usize) -> BatchExecuteResult {
+        assert!(!self.is_ended);
+
+        if self.n == 0 {
+            self.is_ended = true;
+            return BatchExecuteResult {
+                data: LazyBatchColumnVec::empty(),
+                warnings: EvalWarnings::default(),
+                is_drained: Ok(true),
+            };
+        }
+
+        let result = self.handle_next_batch();
+
+        match result {
+            Err(e) => {
+                // When there are error, we can just return empty data.
+                self.is_ended = true;
+                BatchExecuteResult {
+                    data: LazyBatchColumnVec::empty(),
+                    warnings: self.context.take_warnings(),
+                    is_drained: Err(e),
+                }
+            }
+            Ok(Some(data)) => {
+                self.is_ended = true;
+                BatchExecuteResult {
+                    data,
+                    warnings: self.context.take_warnings(),
+                    is_drained: Ok(true),
+                }
+            }
+            Ok(None) => BatchExecuteResult {
+                data: LazyBatchColumnVec::empty(),
+                warnings: self.context.take_warnings(),
+                is_drained: Ok(false),
+            },
+        }
+    }
+
+    #[inline]
+    fn collect_statistics(&mut self, destination: &mut BatchExecuteStatistics) {
+        self.src.collect_statistics(destination);
+    }
+}
+
+/// The item in the heap of `BatchTopNExecutor`.
+///
+/// WARN: The content of this structure is valid only if `BatchTopNExecutor` is valid (i.e.
+/// not dropped). Thus it is called unsafe.
+pub struct HeapItemUnsafe {
+    /// A pointer to the `order_is_desc` field in `BatchTopNExecutor`.
+    order_is_desc_ptr: NonNull<Pin<Vec<bool>>>,
```

Comment:
We shouldn't point to the `order_is_desc` field in `BatchTopNExecutor`. A move of `BatchTopNExecutor` will invalidate it. We can point to the inner slice of the vec instead. That address is stable.

Reply:
BTW, for those fields that don't change after the executor is created, we can simply store a `Box<[T]>` in the executor, it will save one word (

Reply:
I think actually order_is_desc is indeed pinned over the inner slice.

`Box<[T]>` is a good suggestion!

Reply:
I don't think so. Pin ensures the inner slice (`Deref::Target`) is pinned, but `Vec` itself is not protected. Illustration: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=f17281079b78b1294e7106a92f767d64

Reply:
Updated, PTAL again.